src/docfx/build/toc/parser/TableOfContentsParser.cs

```diff
@@ -21,13 +21,48 @@ internal static class TableOfContentsParser
-            Load(Context context, Document file, MonikerComparer monikerComparer, MonikerMap monikerMap, ResolveContent resolveContent, ResolveHref resolveHref, ResolveXref resolveXref)
+        public static (List<Error> errors, List<TableOfContentsItem> items, TableOfContentsMetadata metadata)
+            Load(Context context, Document file, MetadataProvider metadataProvider, MonikerProvider monikerProvider, MonikerMap monikerMap, ResolveContent resolveContent, ResolveHref resolveHref, ResolveXref resolveXref)
         {
-            var (errors, inputModel) = LoadInputModelItems(context, file, file, monikerComparer, monikerMap, resolveContent, resolveHref, resolveXref, new List<Document>());
+            var (errors, inputModel) = LoadInputModel(context, file, file, monikerMap, resolveContent, resolveHref, resolveXref, new List<Document>());
+
+            TableOfContentsMetadata tocMetadata = null;
+            if (metadataProvider != null && monikerProvider != null)
+            {
+                Error monikerError;
+                var (metadataErrors, metadata) = metadataProvider.GetMetadata(file, inputModel.Metadata);
+                errors.AddRange(metadataErrors);
+
+                tocMetadata = metadata.ToObject<TableOfContentsMetadata>();
+                (monikerError, tocMetadata.Monikers) = monikerProvider.GetFileLevelMonikers(file, tocMetadata.MonikerRange);
+                errors.AddIfNotNull(monikerError);
+
+                ProcessTocItemMoniker(inputModel?.Items, tocMetadata.Monikers, monikerProvider.Comparer);
+            }
 
             var items = inputModel?.Items?.Select(r => TableOfContentsInputItem.ToTableOfContentsModel(r)).ToList();
-            return (errors, items, inputModel?.Metadata);
+            return (errors, items, tocMetadata);
+        }
+
+        private static void ProcessTocItemMoniker(List<TableOfContentsInputItem> tocModelItems, List<string> fileMonikers, MonikerComparer comparer)
+        {
+            if (tocModelItems != null)
+            {
+                foreach (var tocModelItem in tocModelItems)
+                {
+                    if (tocModelItem.Items != null)
+                    {
+                        ProcessTocItemMoniker(tocModelItem.Items, fileMonikers, comparer);
+                    }
+                    if (tocModelItem.Href != null && GetHrefType(tocModelItem.Href) == TocHrefType.AbsolutePath)
+                    {
+                        tocModelItem.Monikers = fileMonikers;
+                    }
+                    var childrenMonikers = tocModelItem.Items?.SelectMany(child => child?.Monikers ?? new List<string>()) ?? new List<string>();
+                    tocModelItem.Monikers = childrenMonikers.Union(tocModelItem.Monikers ?? new List<string>()).Distinct(comparer).ToList();
+                    tocModelItem.Monikers.Sort(comparer);
+                }
+            }
         }
 
```

Comment:
suggest to move outside of parser, to make parser clear and clean, focus on model/synatx parsing.

Reply:
In my opinion, since the output of parser contains monikers info, I want it to be correct, but not an intermediate results.
@superyyrrzz @yufeih @live1206 what do you think?

Reply:
In my opinion, since the output of parser contains monikers info, I want it to be correct, but not an intermediate results.
@superyyrrzz @yufeih @live1206 what do you think?

---
In reply to: [243163503](https://github.com/dotnet/docfx/pull/3797#discussion_r243163503) [](ancestors = 243163503)

Reply:
Prefer *moving all monikers related logic out of parser*, over *resolving monikers along with parsing toc item in a centralized place*, over *resolving monikers along with parsing toc item + patching later*.